### PR TITLE
Publish process-completed events

### PR DESCRIPTION
**Pull request by BjarkeMeier**
## Description

More unit tests will be added in an upcoming PR.

## References

* #33

### CODE REVIEW

File: source/dotnet/Application/Batches/BatchApplicationService.cs [hunk 1]
```diff
@@ -22,17 +23,52 @@ public class BatchApplicationService : IBatchApplicationService
...
+        var batch = new Batch(batchRequestDto.ProcessType, gridAreaCodes);
+
         await _batchRepository.AddAsync(batch).ConfigureAwait(false);
         await _unitOfWork.CommitAsync().ConfigureAwait(false);
     }
+
+    public async Task StartRequestedAsync()
+    {
+        var requestedBatches = await _batchRepository.GetRequestedAsync().ConfigureAwait(false);
+        await _batchRunner.BeginExecuteAsync(requestedBatches).ConfigureAwait(false);
```
Comment (Sondergaard) [inline source/dotnet/Application/Batches/BatchApplicationService.cs:49]: Case: requestedBatches contains 7 pending batches. BatchRunner is able to start the first 4 batches and then fails with an exception. How can the application recover from this invalid state where 4 jobs have started but the state is not reflected?

File: source/dotnet/Domain/BatchAggregate/BatchExecutionState.cs [hunk 2]
```diff
@@ -14,8 +14,23 @@
...
+
+    /// <summary>
+    /// The batch is currently executing.
+    /// </summary>
+    Executing = 1,
+
+    /// <summary>
+    /// The batch has (successfully) completed.
+    /// </summary>
+    Completed = 2,
```
Comment (Sondergaard) [inline source/dotnet/Domain/BatchAggregate/BatchExecutionState.cs:35]: Failure is not an option?

Comment (BjarkeMeier) [reply source/dotnet/Domain/BatchAggregate/BatchExecutionState.cs:35]: Eventually yes, but not now.

File: source/dotnet/Application/Batches/BatchApplicationService.cs (hunk 1 above)
Comment (BjarkeMeier) [reply source/dotnet/Application/Batches/BatchApplicationService.cs:49]: I anticipate that when getting there we might do something like this:
Let `BatchRunner.BeginExecuteAsync()` first check if the batches have actually already been started (and possibly finished). And in that case, silently ignore them. Then they will get the "executing" state and eventually be found completed.

Do you have different thoughts?

File: source/dotnet/Domain/BatchAggregate/BatchExecutionState.cs (hunk 2 above)
Comment (Sondergaard) [reply source/dotnet/Domain/BatchAggregate/BatchExecutionState.cs:35]: 👍

File: source/dotnet/Application/Batches/BatchApplicationService.cs (hunk 1 above)
Comment (Sondergaard) [reply source/dotnet/Application/Batches/BatchApplicationService.cs:47]: I would try to break it down and process each batch individually. Handle each batch in it's own unit-of-work.

File: source/dotnet/Application/Batches/BatchApplicationService.cs [hunk 3]
```diff
@@ -22,17 +22,50 @@ public class BatchApplicationService : IBatchApplicationService
...
+
+    public async Task StartPendingAsync()
+    {
+        var batches = await _batchRepository.GetPendingAsync().ConfigureAwait(false);
+
+        // For now we simulate that batches completes immediately after being started
+        batches.ForEach(b => b.SetExecuting());
+        batches.ForEach(b => b.Complete());
+        var completedProcesses = CreateProcessCompletedEvents(batches);
+        await _processCompletedPublisher.PublishAsync(completedProcesses).ConfigureAwait(false);
```
Comment (defectiveAi) [inline source/dotnet/Application/Batches/BatchApplicationService.cs:52]: If Publish fails for some events (e.g. last event gets a timeout, for instance), all batch aggregates will be rolled back. We now have some CompletedEvents that reference an uncompleted batch.

If ServiceBusSender sends all the messages as a batch, you can ignore this comment.

File: source/dotnet/Infrastructure/ProcessCompletedPublisher.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+using Azure.Messaging.ServiceBus;
+using Energinet.DataHub.Wholesale.Application.Processes;
+using Energinet.DataHub.Wholesale.Infrastructure.ServiceBus;
+
+namespace Energinet.DataHub.Wholesale.Infrastructure;
+
+public class ProcessCompletedPublisher : IProcessCompletedPublisher
+{
+    private readonly ServiceBusSender _serviceBusSender;
+    private readonly IServiceBusMessageFactory _serviceBusMessageFactor;
```
Comment (defectiveAi) [inline source/dotnet/Infrastructure/ProcessCompletedPublisher.cs:24]: nitpick: Factor_y_

File: source/dotnet/Infrastructure/Registration/TargetedSingleton.cs [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+namespace Energinet.DataHub.Wholesale.Infrastructure.Registration;
+
+/// <summary>
+/// Allow different instances of same type to be registered as singletons and at
+/// the same time allowed to be distinguished at injection time depending on the dependent.
+/// </summary>
+/// <param name="Instance"></param>
+/// <typeparam name="TInstance">The type of the instance.</typeparam>
+/// <typeparam name="TDependent">The type of the class depending on the instance.</typeparam>
+public sealed record TargetedSingleton<TInstance, TDependent>(TInstance Instance);
```
Comment (defectiveAi) [inline source/dotnet/Infrastructure/Registration/TargetedSingleton.cs:24]: Clever :D

File: source/dotnet/Infrastructure/ProcessCompletedPublisher.cs (hunk 4 above)
Comment (BjarkeMeier) [reply source/dotnet/Infrastructure/ProcessCompletedPublisher.cs]: Thx. Fixed.

File: source/dotnet/Application/Batches/BatchApplicationService.cs (hunk 3 above)
Comment (BjarkeMeier) [reply source/dotnet/Application/Batches/BatchApplicationService.cs:52]: As far as I understand the docs I anticipate that the batch sending is transactional.

File: source/dotnet/Tests/Domain/BatchAggregate/BatchBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+using Energinet.DataHub.Wholesale.Domain.GridAreaAggregate;
+
+namespace Energinet.DataHub.Wholesale.Tests.Domain.BatchAggregate;
+
+public class BatchBuilder
+{
+    private BatchExecutionState? _state;
+    private List<GridAreaCode> _gridAreaCodes = new() { new("805") };
+
+    public BatchBuilder WithState(BatchExecutionState state)
```
Comment (defectiveAi) [inline source/dotnet/Tests/Domain/BatchAggregate/BatchBuilder.cs:26]: Is the state for a new built object not always Pending? Should it be allowed to build a new aggregate with a freely chosen state?

**Review by defectiveAi — APPROVED** [review]

File: source/dotnet/Application/Batches/BatchApplicationService.cs (hunk 1 above)
Comment (BjarkeMeier) [reply source/dotnet/Application/Batches/BatchApplicationService.cs]: The offending code has been removed.

**Pull request merged**